src/meepgeom.cpp

```diff
@@ -364,11 +390,28 @@ meep::vec material_grid_grad(vector3 p, material_data *md) {
 
 #undef D
 
-  gradient.set_direction(meep::X, du_dx);
-  gradient.set_direction(meep::Y, du_dy);
-  gradient.set_direction(meep::Z, du_dz);
+  // [du_dx,du_dy,du_dz] is the gradient ∇u with respect to the transformed coordinate
+  // r1 of the matgrid_val function but what we want is the gradient of u(g(r2)) with 
+  // respect to r2 where g(r2) is the to_geom_object_coords function. computing this
+  // quantity involves using the chain rule and the vector-Jacobian product ∇u J
+  // where J is the Jacobian matrix of g.
+  vector3 grad_u;
+  grad_u.x = du_dx * nx;
+  grad_u.y = du_dy * ny;
+  grad_u.z = du_dz * nz;
+  if (o != NULL) {
+    vector3 grad_u_J = to_geom_object_coords_VJP(grad_u, o);
+    gradient.set_direction(meep::X, grad_u_J.x);
+    gradient.set_direction(meep::Y, grad_u_J.y);
+    gradient.set_direction(meep::Z, grad_u_J.z);
+  }
+  else {
+    gradient.set_direction(meep::X, grad_u.x);
+    gradient.set_direction(meep::Y, grad_u.y);
+    gradient.set_direction(meep::Z, grad_u.z);
```

Comment:
If the material grid is the default material, isn't there still a stretch factor?

Reply:
In particular, you need the chain rule for the `map_lattice_coordinates` function (called from [this line](https://github.com/NanoComp/meep/blob/84162fcead3f38880a1a5a9a53240c39ac3fa255/src/meepgeom.cpp#L447), defined [here](https://github.com/NanoComp/meep/blob/84162fcead3f38880a1a5a9a53240c39ac3fa255/src/meepgeom.cpp#L417-L424).

Reply:
So you just need to divide by `geometry_lattice.size.x` etc.

Reply:
In particular, use `geometry_lattice.size.x == 0 ? 0 : grad_u.x / geometry_lattice.size.x` etc.